relayer/src/link.rs

```diff
@@ -1426,34 +1429,42 @@ impl RelayPath {
         Ok(())
     }
 
-    /// Pulls out the operational elements with elapsed delay period and that can
+    /// Destination-chain fetching of operational data.
+    ///
+    /// Pulls out an operational elements with elapsed delay period and that can
     /// now be processed. Does not block: if no OD fulfilled the delay period (or none is
-    /// scheduled), returns immediately with `vec![]`.
-    fn try_fetch_scheduled_operational_data(
-        &mut self,
-    ) -> (Vec<OperationalData>, Vec<OperationalData>) {
-        // The first elements of the op. data vector contain the oldest entry.
-        // Remove and return the elements with elapsed delay.
-
-        let src_ods: Vec<OperationalData> = self
-            .src_operational_data
+    /// scheduled), returns immediately with `None`.
+    fn try_fetch_scheduled_operational_data_dst(&mut self) -> Option<OperationalData> {
+        // Check if the delay elapsed for some item, and return its position.
+        let found = self
+            .dst_operational_data
             .iter()
-            .filter(|op| op.scheduled_time.elapsed() > self.channel.connection_delay)
-            .cloned()
-            .collect();
+            .position(|op| op.scheduled_time.elapsed() > self.channel.connection_delay);
 
-        self.src_operational_data = self.src_operational_data[src_ods.len()..].to_owned();
+        if let Some(pos) = found {
+            let item = self.dst_operational_data.remove(pos);
+            Some(item)
+        } else {
+            None
+        }
+    }
 
-        let dst_ods: Vec<OperationalData> = self
-            .dst_operational_data
+    /// Source-chain fetching of operational data.
+    ///
+    /// Equivalent to [`try_fetch_scheduled_operational_data_dst`]
+    /// but for the source chain operational data.
+    fn try_fetch_scheduled_operational_data_src(&mut self) -> Option<OperationalData> {
```

Comment:
We no longer return a vector of op. data, but at most one.

The rationale is as follows: Returning a vector was problematic because one of them might fail to process, and then Hermes would never get a chance to process the subsequent ones.